sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,912 @@
+            }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_from_dec_str(
+                    in_dec_str_ptr: *const u64,
+                    in_size: u64,
+                    out_ptr: *mut u8,
+                    out_size_ptr: *mut u64,
+                    out_is_neg_flag_ptr: *mut u64,
+                ) -> u64;
+            }
+            let mut value = Vec::<u8>::with_capacity(string.len());
+            let mut value_len = string.len() as u64;
+            let mut is_negative = 0u64;
+            unsafe {
+                let result = sol_bignum_from_dec_str(
+                    string.as_ptr() as *const _ as *const u64,
+                    string.len() as u64,
+                    value.as_mut_ptr() as *mut _ as *mut u8,
+                    &mut value_len as *mut _ as *mut u64,
+                    &mut is_negative as *mut _ as *mut u64,
+                );
+                if result == 0 {
+                    value.set_len(value_len as usize);
+                    Ok(Self {
+                        is_negative: is_negative != 0,
+                        value,
+                    })
+                } else {
+                    Err(result)
+                }
+            }
+        }
+    }
+
+    /// Flag indicating if negative or not
+    pub fn is_negative(&self) -> bool {
+        self.is_negative
+    }
+
+    /// Returns an array of bytes of self
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.value
+    }
+
+    /// Add BigNumbers
+    pub fn add(&self, rhs: &BigNumber) -> Result<Self, u64> {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            match openssl::bn::BigNum::from_slice(&self.value) {
+                Ok(my_num) => match openssl::bn::BigNum::from_slice(&rhs.value) {
+                    Ok(rhs) => match &mut openssl::bn::BigNum::new() {
+                        Ok(bn_done) => match bn_done.checked_add(&my_num, &rhs) {
+                            Ok(_) => Ok(Self {
+                                is_negative: bn_done.is_negative(),
+                                value: bn_done.to_vec(),
+                            }),
+                            Err(_) => Err(0x1),
+                        },
+                        Err(_) => Err(0x1),
+                    },
+                    Err(_) => Err(0x1),
+                },
+                Err(_) => Err(0x1),
+            }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_add(
+                    arg_address: *const u64,
+                    arg_count: u64,
+                    out_value_address: *mut u8,
+                    out_neg_address: *mut u64,
+                    out_size_address: *mut u64,
+                ) -> u64;
+            }
+            // Setup the argument array (3)
+            let neg_values = vec![self.is_negative as u8, rhs.is_negative as u8];
```

Comment:
Instead of these heap allocations, you could allocate a structure on the stack that includes a pointer to the value and is_negative and pass that across the syscall (ffi) boundary.  This would require a double translate in the syscall implementation one for the FFI struct and another for the value pointer but that would be cheaper then allocating from the heap in-program.

```
struct FfiBigNumber {
   pub value: &[u8],
   pub is_negative: u8,
}
```

Reply:
So just then transfer the resulting FfiBigNumber values and ownership to the program BigNumber returned to the caller?

Reply:
FFIBigNumber would only be for sending it across the syscall (ffi) boundary

Reply:
Got it, implemented